crmsh/ui_cluster.py

```diff
@@ -91,78 +138,94 @@ def __init__(self):
         self._inventory_target = None
 
     @command.skill_level('administrator')
-    def do_start(self, context):
+    def do_start(self, context, *args):
         '''
-        Starts the cluster services on this node
+        Starts the cluster services on all nodes or specific node(s)
         '''
-        try:
-            if utils.service_is_active("pacemaker.service"):
-                logger.info("Cluster services already started")
-                return
-            bootstrap.start_pacemaker()
-            if utils.is_qdevice_configured():
-                utils.start_service("corosync-qdevice")
-            logger.info("Cluster services started")
-        except IOError as err:
-            context.fatal_error(str(err))
-
-        # TODO: optionally start services on all nodes or specific node
+        node_list = parse_option_for_nodes(context, *args)
+        for node in node_list[:]:
+            if utils.service_is_active("pacemaker.service", remote_addr=node):
+                logger.info("Cluster services already started on {}".format(node))
+                node_list.remove(node)
+        if not node_list:
+            return
+
+        bootstrap.start_pacemaker(node_list)
+        if utils.is_qdevice_configured():
+            utils.start_service("corosync-qdevice", node_list=node_list)
+        for node in node_list:
+            logger.info("Cluster services started on {}".format(node))
 
     @command.skill_level('administrator')
-    def do_stop(self, context):
+    def do_stop(self, context, *args):
         '''
-        Stops the cluster services on this node
+        Stops the cluster services on all nodes or specific node(s)
         '''
-        try:
-            if not utils.service_is_active("corosync.service"):
-                logger.info("Cluster services already stopped")
-                return
-            if utils.service_is_active("corosync-qdevice"):
-                utils.stop_service("corosync-qdevice")
-            utils.stop_service("corosync")
-            logger.info("Cluster services stopped")
-        except IOError as err:
-            context.fatal_error(str(err))
-
-        # TODO: optionally stop services on all nodes or specific node
+        node_list = parse_option_for_nodes(context, *args)
+        for node in node_list[:]:
+            if not utils.service_is_active("corosync.service", remote_addr=node):
+                if utils.service_is_active("sbd.service", remote_addr=node):
+                    utils.stop_service("corosync", remote_addr=node)
+                    logger.info("Cluster services stopped on {}".format(node))
+                else:
+                    logger.info("Cluster services already stopped on {}".format(node))
+                node_list.remove(node)
+        if not node_list:
+            return
+
+        bootstrap.wait_for_cluster("Waiting for {} online".format(' '.join(node_list)), node_list)
```

Comment:
Yes.

- Just after `crm cluster start --all`, the status of all nodes are `UNCLEAN`, on this time, if execute `crm cluster stop --all`, I found the stop process will be hang, and on some node, the output of `crm_mon` will show all nodes' status are `pending`
- If all nodes already `Online`, in `wait_for_cluster` function, I made a check firstly, and the line `Waiting for .... online` will be quiet

Reply:
Good point. 

I can understand the wait experience in bootstrap. Well, probably no necessary to force user to wait `crm cluster start`.

I would rather think to remove wait in `crm cluster start`, and let `crm cluster stop` directly abort if the criteria not meet. It's a reasonable error handling, rather than have sysadmin really wait in front of the screen. BTW, not sure if you want to fix it in this PR, or track this as an issue and fix it later on.

For those scripts do want to 'wait', before we implement '--wait' option, the following example steps could help
crm cluster start
crm wait_for_startup
crm cluster stop